orttraining/orttraining/training_ops/cuda/math/scaled_sum_impl.cu

```diff
@@ -0,0 +1,166 @@
+          if (InputCount == 3)
+            output_value[i] = input_values[0][i] * static_cast<T>(scales_[0]) +
+                              input_values[1][i] * static_cast<T>(scales_[1]) +
+                              input_values[2][i] * static_cast<T>(scales_[2]);
+          else
+            output_value[i] = input_values[0][i] * static_cast<T>(scales_[0]) +
+                              input_values[1][i] * static_cast<T>(scales_[1]);
+        }
+      }
+
+      *reinterpret_cast<LoadT*>(&output_data_[id]) = *reinterpret_cast<LoadT*>(&output_value[0]);
+    } else {
+      T* output_value = output_data_ + id;
+#pragma unroll
+      for (int i = 0; i < NumUnroll; i++) {
+        CUDA_LONG li = id + i;
+        if (li < N_) {
+          if (InputCount == 3)
+            output_value[i] = input_values[0][i] * static_cast<T>(scales_[0]) +
+                              input_values[1][i] * static_cast<T>(scales_[1]) +
+                              input_values[2][i] * static_cast<T>(scales_[2]);
+
+          else
+            output_value[i] = input_values[0][i] * static_cast<T>(scales_[0]) +
+                              input_values[1][i] * static_cast<T>(scales_[1]);
+        }
+      }
+    }
+  }
+
+ private:
+  const T* inputs_[InputCount];
+  float scales_[InputCount];
+  CUDA_LONG N_;
+  T* output_data_;
+};
+
+template <typename FuncT>
+__global__ void ScaledSumKernel(const FuncT functor) {
+  CUDA_LONG idx = blockDim.x * blockIdx.x + threadIdx.x;
+  functor(idx);
+}
+
+template <typename T>
+void ScaledSumImpl(cudaStream_t stream,
+                   int64_t input_element_count,
+                   const std::vector<const T*>& inputs,
+                   const std::vector<float>& scales,
+                   T* output_data) {
+  const int blocksPerGrid = static_cast<int>(CeilDiv(input_element_count, kBlockSize * kNumUnroll));
+
+  const bool use_vectorized = (input_element_count % kNumUnroll == 0) &&
+                              (reinterpret_cast<uintptr_t>(output_data) % 32 == 0) &&
+                              (reinterpret_cast<uintptr_t>(inputs[0]) % 32 == 0) &&
+                              (reinterpret_cast<uintptr_t>(inputs[1]) % 32 == 0) &&
+                              (inputs.size() < 3 || (reinterpret_cast<uintptr_t>(inputs[2]) % 32 == 0));
+
+  const int input_count = static_cast<int>(inputs.size());
+  using TwoInputTVectorizedFunctorType = ScaledSumFunctor<T, kNumUnroll, 2, true>;
+  using TwoInputTNonVectorizedFunctorType = ScaledSumFunctor<T, kNumUnroll, 2, false>;
+  using ThreeInputTVectorizedFunctorType = ScaledSumFunctor<T, kNumUnroll, 3, true>;
+  using ThreeInputTNonVectorizedFunctorType = ScaledSumFunctor<T, kNumUnroll, 3, false>;
+
+  if (input_count == 2) {
+    if (use_vectorized) {
+      auto functor = TwoInputTVectorizedFunctorType(inputs, scales, input_element_count, output_data);
+      ScaledSumKernel<TwoInputTVectorizedFunctorType><<<blocksPerGrid, kBlockSize, 0, stream>>>(functor);
```

Comment:
why no `ScaledSumKernel<TwoInputTVectorizedFunctorType><<<blocksPerGrid, kBlockSize, 0, stream>>>(TwoInputTVectorizedFunctorType(inputs, scales, input_element_count, output_data));` ?

Reply:
I used this for debugging earlier, changed it back to align with others.